Reconstruction/Digitisers/src/G2CDArborAlg.cpp

```diff
@@ -0,0 +1,976 @@
+     int DHIndexI = 0;
+     int DHIndexJ = 0; 
+     float DHChargeWeight = 0;
+     int DHCellID0 = 0; 
+     float RndCharge = 0;
+     int SingleMCPPID = 0; 
+     //float SingleMCPPEn = 0; 
+     float RefPosX = 0;
+     float RefPosY = 0;
+     float RefPosZ = 0;
+     float DeltaPosI = 0;
+     float DeltaPosJ = 0;
+     std::vector<float> CurrWeightVec;			
+     float WeiI = 0; 
+     float WeiJ = 0;
+     int DeltaI = 0;
+     int DeltaJ = 0; 
+     int MapI = 0; 
+     int MapJ = 0; 
+     int MapIndex = 0; 
+     float FHitPos[3] = {0, 0, 0};
+     float currTime = 0;
+     float HitStepEn = 0;
+     float EmaxStep = 0;
+
+     // LCCollectionVec *relcol = new LCCollectionVec(LCIO::LCRELATION);
+     // LCFlagImpl linkflag; 
+     // linkflag.setBit(LCIO::CHBIT_LONG);
+     // relcol->setFlag(linkflag.getFlag());	
+     edm4hep::MCRecoCaloAssociationCollection* relcol = _caloTruthLinkCollection.createAndPut();
+
+     // LCCollection * MCPCol = evtP->getCollection("MCParticle");
+     // for ( int s0(0); s0 < MCPCol->getNumberOfElements(); s0++)
+     // {
+     // 	    MCParticle * a_MCP = dynamic_cast<MCParticle*> (MCPCol->getElementAt(s0));
+     // 	    if( a_MCP->getParents().size() == 0 )
+     // 	    {
+     // 		 SingleMCPPID = a_MCP->getPDG();
+     // 		 //SingleMCPPEn = a_MCP->getEnergy();
+     // 		 break; 
+     // 	    }
+     // }
+
+     // LCFlagImpl flag;
+     // flag.setBit(LCIO::CHBIT_LONG);                  //To set position & ID1
+     // flag.setBit(LCIO::CHBIT_ID1);
+     // flag.setBit(LCIO::RCHBIT_ENERGY_ERROR); //In order to use an additional FLOAT
+     // flag.setBit(LCIO::RCHBIT_TIME);
+
+
+     double totalEnergy = 0.;
+     for (unsigned int k0 = 0; k0 < _ecalCollections.size(); ++k0)
+     {
+     	  //   // try{
+	  // LCCollection *Ecalcol = evtP->getCollection( _ecalCollections[k0].c_str() ) ;
+	  // CellIDDecoder<SimCalorimeterHit> idDecoder( Ecalcol );
+	  // int NumEcalhit = Ecalcol->getNumberOfElements();
+	  // LCCollectionVec *ecalcol = new LCCollectionVec(LCIO::CALORIMETERHIT);
+	  // ecalcol->setFlag(flag.getFlag());
+	  // string EcalinitString = Ecalcol->getParameters().getStringVal(LCIO::CellIDEncoding);
+	  // ecalcol->parameters().setValue(LCIO::CellIDEncoding, EcalinitString);	
+
+	  // for(int k1 = 0; k1 < NumEcalhit; k1++)
+	  // {	
+
+	  edm4hep::CalorimeterHitCollection* ecalcol = _outputEcalCollections[k0]->createAndPut();
+	  auto Ecalcol = _ecalCollections[k0]->get();
+	  for (auto SimEcalhit: *Ecalcol){
+
+	       // SimCalorimeterHit * SimEcalhit = dynamic_cast<SimCalorimeterHit*>( Ecalcol->getElementAt( k1 ) ) ;
+	       // HitEn = SimEcalhit->getEnergy();
+	       // LayerNum = idDecoder(SimEcalhit)["K-1"];
+	       string m_encoder_str = "M:3,S-1:3,I:9,J:9,K-1:6";
+	       // edm4hep::SimCalorimeterHit aa(SimEcalhit.getCellID(), SimEcalhit.getEnergy(), SimEcalhit.getPosition());
+	       ID_UTIL::CellIDDecoder<edm4hep::SimCalorimeterHit> cellIdDecoder(m_encoder_str);
```

Comment:
As the decoder always uses the same encoder string, it is better to instance once and move the constructor out of the loop.